core/trino-main/src/test/java/io/trino/sql/query/QueryAssertions.java

```diff
@@ -323,30 +326,54 @@ private QueryAssert(
             this.skipResultsCorrectnessCheckForPushdown = skipResultsCorrectnessCheckForPushdown;
         }
 
-        // TODO for better readability, replace this with `exceptColumns(String... columnNamesToExclude)` leveraging MaterializedResult.getColumnNames
-        @Deprecated
-        public QueryAssert projected(int... columns)
+        public QueryAssert exceptColumns(String... columnNamesToExclude)
         {
+            validateIfColumnsPresent(columnNamesToExclude);
+            checkArgument(columnNamesToExclude.length < actual.getColumnNames().size(), "All columns cannot be excluded");
+            return projected(((Predicate<String>) Arrays.asList(columnNamesToExclude)::contains).negate());
+        }
+
+        public QueryAssert projected(String... columnNamesToInclude)
+        {
+            validateIfColumnsPresent(columnNamesToInclude);
+            checkArgument(columnNamesToInclude.length > 0, "At least one column must be projected");
+            return projected(Arrays.asList(columnNamesToInclude)::contains);
+        }
+
+        private QueryAssert projected(Predicate<String> columnFilter)
+        {
+            List<String> columnNames = actual.getColumnNames();
+            Map<Integer, String> columnsIndexToNameMap = columnNames.stream()
+                    .filter(columnFilter)
+                    .collect(toImmutableMap(columnNames::indexOf, Function.identity()));
```

Comment:
columnNames::indexOf is quadratic

use ordinary indexed loop to populate columnsIndexToNameMap